backend/data/src/main/kotlin/io/tolgee/service/security/AuthProviderChangeService.kt

```diff
@@ -1,117 +1,269 @@
 import io.tolgee.util.addMinutes
 import org.apache.commons.lang3.time.DateUtils
 import org.springframework.context.annotation.Lazy
 import org.springframework.stereotype.Service
 import org.springframework.transaction.annotation.Propagation
 import org.springframework.transaction.annotation.Transactional
 import java.util.Calendar
+import java.util.UUID
 
 @Service
 class AuthProviderChangeService(
+  private val tolgeeProperties: TolgeeProperties,
   private val authProviderChangeRequestRepository: AuthProviderChangeRequestRepository,
   private val tenantService: TenantService,
+  private val userAccountService: UserAccountService,
+  private val organizationService: OrganizationService,
   private val organizationRoleService: OrganizationRoleService,
   private val currentDateProvider: CurrentDateProvider,
+  private val authenticationFacade: AuthenticationFacade,
   @Suppress("SelfReferenceConstructorParameter") @Lazy
   private val self: AuthProviderChangeService,
 ) {
   fun getCurrent(user: UserAccount): AuthProviderDto? {
-    return user.asAuthProviderDto()
+    return user.asAuthProviderDto(tolgeeProperties)
   }
 
   fun getRequestedChange(userAccount: UserAccount): AuthProviderDto? {
     return getActiveAuthProviderChangeRequest(userAccount)?.asAuthProviderDto()
   }
 
-  fun initiateProviderChange(data: AuthProviderChangeData) {
-    if (data.accountType == UserAccount.AccountType.MANAGED) {
-      self.saveProviderChange(data)
-      throw AuthenticationException(Message.THIRD_PARTY_SWITCH_INITIATED)
+  fun tryInitiate(
+    resolvedUser: UserAccount?,
+    userEmail: String,
+    data: AuthProviderChangeData,
+  ) {
+    val currentUser = authenticationFacade.authenticatedUserOrNull
+
+    if (currentUser == null) {
+      // Provider change can be only initiated for already signed-in user
+      return
+    }
+
+    if (resolvedUser?.id == currentUser.id) {
+      // Trying to log-in as already logged-in user??
+      return
     }
 
-    // Changing authentication provider to non-SSO providers is disabled for now
-    throw AuthenticationException(Message.USERNAME_ALREADY_EXISTS)
+    val targetUser = resolvedUser ?: userAccountService.findActive(userEmail)
+    if (targetUser?.id != currentUser.id) {
+      // User is trying to switch third party authentication provider, but
+      // the account e-mail does not match the email of the third party account
+      throw AuthenticationException(Message.THIRD_PARTY_SWITCH_CONFLICT)
+    }
+
+    self.forceSave(targetUser, data)
+    throw AuthenticationException(Message.THIRD_PARTY_SWITCH_INITIATED)
   }
 
   @Transactional(propagation = Propagation.REQUIRES_NEW)
-  fun saveProviderChange(data: AuthProviderChangeData) {
-    val req =
-      authProviderChangeRequestRepository.findByUserAccountId(data.userAccount.id)
-        ?: AuthProviderChangeRequest()
-    authProviderChangeRequestRepository.save(req from data)
+  protected fun forceSave(
```

Comment:
Name it `saveInNewTransaction` instead and explain why you're doing so in the comment.